### PR TITLE
KEYCLOAK-19699 RSA key provider with key use = enc cannot select corresponding algorithm on Admin Console

### PR DESCRIPTION
**Pull request by tnorimat**
This PR is for fixing a bug reported by [keycloak-dev](https://groups.google.com/g/keycloak-dev/c/BlHeL4AIO80/m/b52VSIOhAgAJ) and  by [KEYCLOAK-19699](https://issues.redhat.com/browse/KEYCLOAK-19699) that describes it in detail.

### CODE REVIEW

File: services/src/main/java/org/keycloak/keys/GeneratedRsaKeyProviderFactory.java [hunk 1]
```diff
@@ -18,136 +18,69 @@
...
  * @author <a href="mailto:sthorger@redhat.com">Stian Thorgersen</a>
  */
-public class GeneratedRsaKeyProviderFactory extends AbstractRsaKeyProviderFactory {
+public class GeneratedRsaKeyProviderFactory extends AbstractGeneratedRsaKeyProviderFactory {
 
     private static final Logger logger = Logger.getLogger(GeneratedRsaKeyProviderFactory.class);
 
     public static final String ID = "rsa-generated";
 
     private static final String HELP_TEXT = "Generates RSA keys and creates a self-signed certificate";
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/keys/GeneratedRsaKeyProviderFactory.java:38]: Should the help text be updated to explicitly mention that it is used for "signatures" ? Ideally align it with the help text you used for new "enc" based provider

**Review by mposolda — CHANGES_REQUESTED** [review]
@tnorimat Thanks Takashi!

I've added one minor comment inline.

One more question: Is this change backwards compatible? I wonder if someone already used `rsa-generated` provider with the ENC use (even if he was not able to select algorithm before). Or wasn't it possible before this PR at all to use `rsa-generated` for the ENC?
If backwards compatibility is an issue, I wonder that possible alternative solution might be instead of introducing new provider, just change the method `isSupportedRsaAlgorithm` to return algorithms based on whether the use is SIG or ENC. But I am not sure... If you think this PR is fine from backwards compatible perspective, I am ok with the approach you used in the PR.

**Comment by mposolda** [comment]
@tnorimat ah, and I see that there are failing tests. This would also require fix...

**Comment by tnorimat** [comment]
@mposolda It seems that test itself have problems. I'll try to fix them.

OIDCAdvancedRequestParamsTest#testRealmPublicKeyEncryptedRequestObjectUsingKid
OIDCAdvancedRequestParamsTest#testRealmPublicKeyEncryptedRequestObjectUsingRSA_OAEPWithA128CBC_HS256
OIDCAdvancedRequestParamsTest#testRealmPublicKeyEncryptedRequestObjectUsingRSA_OAEP_256WithA256GCM
OIDCAdvancedRequestParamsTest#testSignedAndEncryptedRequestObject
OIDCAdvancedRequestParamsTest#testWrongContentEncryptionAlgorithm
OIDCAdvancedRequestParamsTest#testWrongEncryptionAlgorithm

**Comment by tnorimat** [comment]
@mposolda Also, I will consider backward compatibility issue and get back to you.

**Comment by tnorimat** [comment]
@mposolda I've concluded that this original PR does not break backward compatibility. The reason is as follows.

- The current keycloak does not look at an algorithm for key provider whose key use is "enc".
The key provider whose key use is "enc" has been used for encrypting client generating JWT (e.g. Request Object).
The current keycloak uses this key provider to decode this encrypted JWT in `DefaultTokenManager.decodeClientJWT` .

https://github.com/keycloak/keycloak/blob/58d403cf24e6d57d2776d6b85f914db3eecf331d/services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java#L133-L140

These codes check key use but do not check key algorithm. Therefore, the existing key provider whose key use is "enc" and key algorithm is like "PS256" still works well.

However, to check this backward compatibility by the following existing tests, I've modified the codes of `GeranatedRsaKeyProviderFactory.create` .

OIDCAdvancedRequestParamsTest#testRealmPublicKeyEncryptedRequestObjectUsingKid
OIDCAdvancedRequestParamsTest#testRealmPublicKeyEncryptedRequestObjectUsingRSA_OAEPWithA128CBC_HS256
OIDCAdvancedRequestParamsTest#testRealmPublicKeyEncryptedRequestObjectUsingRSA_OAEP_256WithA256GCM
OIDCAdvancedRequestParamsTest#testSignedAndEncryptedRequestObject
OIDCAdvancedRequestParamsTest#testWrongContentEncryptionAlgorithm
OIDCAdvancedRequestParamsTest#testWrongEncryptionAlgorithm

**Comment by tnorimat** [comment]
There are two options to implement this feature.
        
1. Add new key provider for "enc"
2. Make existing key provider show algirithms depending on its key use
        
I've taken the fist option by the following reasons.
        
 - By leveraging angularJS's feature, the existing key provider's configuration UI is automatically rendered without its own javascript codes/HTML/FTL. I'm afraid that if I took the second option, I would have to add new javascript logic and FTL/HTML.
        
- All other key provider does not allow keycloak admin to select key use.
          aes-generated(enc), ecdsa-generated(sig), hmac-generated(sig), java-keystore(sig), rsa(sig) 
I've followed this style.
          rsa-generated(sig), rsa-enc-generated(enc)

File: services/src/main/java/org/keycloak/keys/GeneratedRsaKeyProviderFactory.java (hunk 1 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/keys/GeneratedRsaKeyProviderFactory.java:38]: @mposolda Yes, I will change the help text as follows.

`"Generates RSA signature keys and creates a self-signed certificate"`

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks! This looks good to me.

I will merge after Keycloak 16 release (AFAIK we try to avoid merge non-critical issues into Keycloak master now to avoid the potential regressions in the Keycloak 16 release).

**Pull request merged**